platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/storage/FileSource.java

```diff
@@ -217,6 +261,95 @@ public static String getInternalCachePath(@NonNull Context context) {
+
+    if (getInstance(context).isActivated()) {
+      final String activatedMessage = "Cannot set path, file source is activated!";
+      Logger.w(TAG, activatedMessage);
+      callback.onError(activatedMessage);
+    } else {
+      final WeakReference<Context> contextWeakReference = new WeakReference<>(context);
+      new Thread(new Runnable() {
+        @Override
+        public void run() {
+          final Context context = contextWeakReference.get();
+          final String message;
+          if (context != null) {
+            if (resourcesCachePath != null) {
+              if (!isExternalStorageConfiguration(context)) {
+                message = "External storage setting is not enabled";
+              } else if (resourcesCachePath.equals(path)) {
+                message = "Changing the path not needed";
+              } else if (!isExternalStorageReadable()) {
+                message = "External storage is not readable";
+              } else if (!isPathWritable(path)) {
+                message = "Path is not writable: " + path;
+              } else {
+                message = null;
+
+                final SharedPreferences.Editor editor =
+                    context.getSharedPreferences(MAPBOX_SHARED_PREFERENCES, Context.MODE_PRIVATE).edit();
+                editor.putString(MAPBOX_SHARED_PREFERENCE_RESOURCES_CACHE_PATH, path).apply();
+
+                new Handler(Looper.getMainLooper()).post(new Runnable() {
+                  @Override
+                  public void run() {
+                    setResourcesCachePath(context, path);
+                    callback.onSuccess(path);
+                  }
+                });
+              }
+            } else {
+              message = "Resources cache path was not initialized.";
+            }
+          } else {
+            message = "Context is null";
+          }
+
+          if (message != null) {
+            Logger.w(TAG, message);
+            new Handler(Looper.getMainLooper()).post(new Runnable() {
+              @Override
+              public void run() {
+                callback.onError(message);
+              }
+            });
+          }
+        }
+      }).start();
+    }
+  }
+
+  private static void setResourcesCachePath(@NonNull Context context, @NonNull String path) {
+    resourcesCachePathLoaderLock.lock();
+    resourcesCachePath = path;
+    reinitializeOfflineManager(context);
+    resourcesCachePathLoaderLock.unlock();
+  }
+
+  private static boolean isPathWritable(String path) {
+    if (path == null || path.isEmpty()) {
+      return false;
+    }
+    return new File(path).canWrite();
+  }
+
+  private static void reinitializeOfflineManager(@NonNull Context context) {
+    final FileSource fileSource = FileSource.getInstance(context);
+    fileSource.initialize(Mapbox.getAccessToken(), resourcesCachePath, context.getResources().getAssets());
```

Comment:
I'm not sure this should be the way to go - with this approach you're recreating core `DefaultFileSource` and information about previously set options like`ResourceTransform`, `ApiBaseUrl` or `AccessToken` will be lost. Also, even though you're forcing the recreation of the `OfflineManager`'s instance whenever requested again if there already are any instances alive, they'll start to point to an invalid native pointer and crash when accessed.

I think that `mbgl::DefaultFileSource` should simply expose a `cachePath` setter that will recreate the database object with the new path. Any thoughts on this @kkaefer?

Reply:
That's why I said that im not sure if this is a clean way.